js/common/bootstrap/modules/buffer.js

```diff
@@ -162,109 +170,99 @@ global.DEFINE_MODULE('buffer', (function () {
+        return target.parent[name];
       }
-
-      // note: the following condition means we'll always create a SlowBuffer
-      // for non-empty lengths. This is required to circumvent using the shared
-      // pool, which is only allocated but never freed (and therefore leaks
-      // memory on shutdown)
-      if (this.length > 0) {
-        // Big buffer, just alloc one.
-        this.parent = new SlowBuffer(this.length);
-        this.offset = 0;
+      return target[name];
+    },
+    set (target, name, value) {
+      if (typeof name === 'string' && name.match(/^\d+$/)) {
+        target.parent[name] = value;
       } else {
-        // Zero-length buffer
-        this.parent = new SlowBuffer(0);
-        this.offset = 0;
+        target[name] = value;
       }
+      return true;
+    }
+  };
 
-      // optimize by branching logic for new allocations
-      if (typeof subject !== 'number') {
-        if (type === 'string') {
-          // We are a string
-          this.length = this.write(subject, 0, encoding);
-        // if subject is buffer then use built-in copy method
-        } else if (Buffer.isBuffer(subject)) {
-          if (subject.parent)
-            subject.parent.copy(this.parent,
-              this.offset,
-              subject.offset,
-              this.length + subject.offset);
-          else
-            subject.copy(this.parent, this.offset, 0, this.length);
-        } else if (isArrayIsh(subject)) {
-          for (var i = 0; i < this.length; i++)
-            this.parent[i + this.offset] = subject[i];
-        }
-      }
+  function createBufferFromString(subject, encoding) {
+    const buffer = Object.create(Buffer.prototype);
+    const length = Buffer.byteLength(subject, encoding);
+    buffer.parent = new SlowBuffer(length);
+    buffer.offset = 0;
+    buffer.length = buffer.parent.write(subject, encoding);
+    return new Proxy(buffer, BUFFER_PROXY_HANDLER);
+  }
+
+  function createBufferFromArrayBuffer(subject, byteOffset = 0, length = subject.byteLength - byteOffset) {
+    const buffer = Object.create(Buffer.prototype);
+    const dv = new DataView(subject);
+    buffer.parent = new SlowBuffer(length);
+    buffer.offset = 0;
+    buffer.length = length;
+    for (let i = 0; i < length; i++) {
+      buffer.parent[i] = dv.getUint8(byteOffset + i);
     }
+    return new Proxy(buffer, BUFFER_PROXY_HANDLER);
+  }
 
+  function createBufferFromByteArray(subject) {
+    const buffer = Object.create(Buffer.prototype);
+    buffer.parent = new SlowBuffer(subject.length);
+    buffer.offset = 0;
+    buffer.length = subject.length;
+    for (let i = 0; i < subject.length; i++) {
+      buffer.parent[i] = subject[i];
+    }
+    return new Proxy(buffer, BUFFER_PROXY_HANDLER);
+  }
 
-    return new Proxy(this, handler);
-  // SlowBuffer.makeFastBuffer(this.parent, this, this.offset, this.length)
+  function createBufferFromBuffer(subject, length = subject.length, offset = 0) {
```

Comment:
This part seems odd after analysis in devel.
All other functions have `(subject, offset, length)`
besides this function.
Also the caller in genReandomBytes seems to send length offset instead of offset length

Reply:
Yupp, the Node version of the constructor when passed a buffer doesn't actually take a length and offset but we do. Sadly we also decided to use the inverse order from how Node does it for cases that do take those arguments, so this one is flipped. As `Buffer.from` previously just used the constructor, I decided to mimic the previous behavior instead of making this consistent (and potentially breaking third-party code that relies on this order).

Not sure what you mean about genRandomBytes, that function only uses a single argument with the constructor (size).

Reply:
Nevermind, read your messages on Slack. I'll look into this.